graphql_compiler/query_pagination/query_parameterizer.py

```diff
@@ -1,61 +1,314 @@
+    return new_ast, new_parameters
+
+
+def _make_binary_filter_directive_node(op_name: str, param_name: str) -> DirectiveNode:
+    """Make a binary filter directive node with the given binary op_name and parameter name."""
+    return DirectiveNode(
+        name=NameNode(value="filter"),
+        arguments=[
+            ArgumentNode(name=NameNode(value="op_name"), value=StringValueNode(value=op_name),),
+            ArgumentNode(
+                name=NameNode(value="value"),
+                value=ListValueNode(values=[StringValueNode(value="$" + param_name)]),
+            ),
+        ],
+    )
+
+
+def generate_parameterized_queries(
+    schema_info: QueryPlanningSchemaInfo,
+    query: ASTWithParameters,
+    vertex_partition: VertexPartitionPlan,
+    parameter_value: Any,
+) -> Tuple[ASTWithParameters, ASTWithParameters]:
     """Generate two parameterized queries that can be used to paginate over a given query.
 
-    In order to paginate arbitrary GraphQL queries, additional filters may need to be added to be
-    able to limit the number of results in the original query. This function creates two new queries
-    with additional filters stored as PaginationFilters with which the query result size can be
-    controlled.
+    The first query is produced by adding a "<" filter to a field in the original, and the
+    second by adding a ">=" filter. The parameter value given is used in this filter. This
+    function will potentially remove any existing filters that are no longer needed after
+    the new filter is inserted.
+
+    If the parameter_value is set such that the newly produced query is equivalent to the
+    original query, an AssertionError is raised. Therefore, the parameter_value should be
+    a value inside the range of initial possible values for that field.
 
     Args:
         schema_info: QueryPlanningSchemaInfo
-        query_ast: Document, query that is being paginated.
-        parameters: dict, list of parameters for the given query.
+        query: the query to parameterize
+        vertex_partition: pagination plan dictating where to insert the filter
+        parameter_value: the value of the parameter used for pagination
 
     Returns:
-        ParameterizedPaginationQueries namedtuple
+        tuple (next_page, remainder)
+        next_page: AST and params for next page.
+        remainder: AST and params for the remainder query that returns all results
+                   not on the next page.
     """
-    raise NotImplementedError()
+    query_string = print_ast(query.query_ast)
+    query_root = get_only_query_definition(query.query_ast, GraphQLError)
+
+    # Create extended parameters that include the pagination parameter value
+    param_name = _generate_new_name("__paged_param", set(query.parameters.keys()))
+    extended_parameters = dict(query.parameters)
+    extended_parameters[param_name] = parameter_value
+
+    next_page_root, next_page_parameters = _add_pagination_filter_recursively(
```

Comment:
Any reason not to push the `ASTWithParameters` creation into this function? It seems like we get a raw tuple out and then immediately turn it into a `ASTWithParameters` at all call sites, as far as I can tell.

Reply:
It's a recursive function. If you make it create a `DocumentNode` on output, you'll get the wrong result. I can wrap it in a helper function if that's what you're suggesting